### PR TITLE
Implementing TabControl!

### PR DESCRIPTION
**Pull request by Darksound9**
Commit name says it all.
Implementing TabControl, 2 MyMods checkbox related bug-fixes, implemented Select All + Select Disabled

### CODE REVIEW

File: Api/Modpacks.cs [hunk 1]
```diff
@@ -28,7 +28,8 @@ static class Modpacks {
         public static void AddRowButton_Click(object sender, EventArgs e) {
             OpenFileDialog ofd = new OpenFileDialog {
                 InitialDirectory = "::{20D04FE0-3AEA-1069-A2D8-08002B30309D}",  // using the GUID to access 'This PC' folder
-                Multiselect = true
+                Multiselect = true,
+                Title = "Select a new mod file to apply to your game:"
```
Comment (executionByFork) [inline Api/Modpacks.cs]: Users can select multiple files in this dialog, but also the create modpack tab doesn't apply anything to the game. I think it's best to reword this to: "Select one or more mod files"

File: Api/Modpacks.cs [hunk 2]
```diff
@@ -357,7 +358,8 @@ static class Modpacks {
 
             if ((string)btn.Tag == "btn1") {
                 OpenFileDialog ofd = new OpenFileDialog {
-                    InitialDirectory = "::{20D04FE0-3AEA-1069-A2D8-08002B30309D}"  // using the GUID to access 'This PC' folder
+                    InitialDirectory = "::{20D04FE0-3AEA-1069-A2D8-08002B30309D}",  // using the GUID to access 'This PC' folder
+                    Title = "Select a new mod file to apply to your game:"
```
Comment (executionByFork) [inline Api/Modpacks.cs]: Similarly to the previous comment, lets change this to "Select a mod file" (since you can only select one here)

File: Api/Modpacks.cs [hunk 3]
```diff
@@ -381,7 +383,8 @@ static class Modpacks {
             } else if ((string)btn.Tag == "btn2") {
                 OpenFileDialog ofd = new OpenFileDialog {
                     CheckFileExists = false,    // allow modpack creators to type in a filename for creating new files
-                    InitialDirectory = Config.MCC_home
+                    InitialDirectory = Config.MCC_home,
+                    Title = "Select a matching file to overwrite in your game:"
```
Comment (executionByFork) [inline Api/Modpacks.cs]: This dialog can also be used to create a new file. It's not very obvious and I plan to add a help tip to explain, but users can type in a file name after navigating to a folder they want a new file to be created in, and type a name in the bar to have the modpack create a new file. Lets change this to "Select a destination"

File: Api/MyMods.cs [hunk 4]
```diff
@@ -16,13 +16,19 @@ class MyMods {
...
+        public static void SelectDisabled_chb_CheckedChanged(object sender, EventArgs e) {
+            ApplyBulkChecks(((CheckBox)sender).Checked, false);
+        }
+
+        public static void SelectAll_chb_CheckedChanged(object sender, EventArgs e) {
+            Program.MasterForm.selectEnabled_chb.Checked = ((CheckBox)sender).Checked;
+            Program.MasterForm.selectDisabled_chb.Checked = ((CheckBox)sender).Checked;
+
+            ApplyBulkChecks(Program.MasterForm.selectEnabled_chb.Checked, true);
+            ApplyBulkChecks(Program.MasterForm.selectDisabled_chb.Checked, false);
```
Comment (executionByFork) [inline Api/MyMods.cs]: Instead of calling `ApplyBulkChecks()` twice, and therefore iterating over every listed modpack row twice, I would just copy the inner forloop from `ApplyBulkChecks()` and simply have it set every checkbox to `((CheckBox)sender).Checked`. This will make the function o(n) instead of o(2n)

File: Api/MyMods.cs [hunk 5]
```diff
@@ -51,13 +57,16 @@ class MyMods {
             foreach (Panel p in modpacksList) {
                 CheckBox chb = (CheckBox)p.GetChildAtPoint(Config.MyModsChbPoint);
                 string modpackname = chb.Text.Replace(Config.dirtyPadding, "");
-                if (chb.Checked && !Config.IsPatched(modpackname)) {
-                    toPatch.Add(chb);
-                    if ((string)p.Tag != Config.GetCurrentBuild()) {
-                        oldModpacks.Add(modpackname);
+                if (chb.Checked) {
```
Comment (executionByFork) [inline Api/MyMods.cs]: This change breaks how the application currently works. The code you removed checks each row, if the light is red and the checkbox is checked, the manager will patch the game (your new code does this). However, if the light is green and the checkbox is unchecked the mod manager should unpatch it (your new code swaps this, unpatching if it is checked).

I don't think it's a good idea to swap the controls on the users like this, people will get annoyed because they are already used to how it works (it's been like this since initial release). If anything is checked, when the patch button is clicked it adds those mods to the game, and makes sure any unchecked mods are removed from the game. Changing this would required making changes elsewhere too for QoL improvements. Currently when the mod manager loads, it auto checks any modpack which is enabled. With this change, a user simply trying to add one mod to their game would check a new box and then suddenly every patched mod would be uninstalled, leaving only the new one. This also creates a toggling issue, where clicking the button twice because the app is slow could cause the manager to install and then immediately uninstall all mods.

Lets reset this back to how it was

File: MasterForm.Designer.cs [hunk 6]
```diff
@@ -129,7 +129,7 @@ partial class MasterForm {
             // 
             this.megaCaution.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
             this.megaCaution.Image = ((System.Drawing.Image)(resources.GetObject("megaCaution.Image")));
-            this.megaCaution.Location = new System.Drawing.Point(220, 4);
+            this.megaCaution.Location = new System.Drawing.Point(217, 4);
```
Comment (executionByFork) [inline MasterForm.Designer.cs]: Be careful when moving UI elements around. The majority of things in this app are located with coordinates, so moving the position of elements could cause the program to stack trace under certain conditions. I was worried about this here with the megaCaution icon, but it looks like I don't use coords to access it, so it's fine. Just keep this in mind for the future.

File: MasterForm.Designer.cs [hunk 7]
```diff
@@ -183,325 +183,304 @@ partial class MasterForm {
...
-            this.homePanel.Location = new System.Drawing.Point(5, 89);
-            this.homePanel.Name = "homePanel";
-            this.homePanel.Size = new System.Drawing.Size(563, 356);
-            this.homePanel.TabIndex = 4;
+            // betterPBar
+            // 
+            this.betterPBar.Location = new System.Drawing.Point(4, 448);
+            this.betterPBar.Name = "betterPBar";
+            this.betterPBar.Size = new System.Drawing.Size(570, 44);
+            this.betterPBar.TabIndex = 10;
```
Comment (executionByFork) [inline MasterForm.Designer.cs]: I'm not sure if the progress bar needs to have a tab index. Isn't this the order in which UI elements are selected when cycling through them via the tab key?

File: MasterForm.Designer.cs [hunk 8]
```diff
@@ -183,325 +183,304 @@ partial class MasterForm {
...
             this.selectEnabled_chb.Text = "Select Enabled";
             this.selectEnabled_chb.UseVisualStyleBackColor = true;
             // 
             // homeEnabledLabel
             // 
             this.homeEnabledLabel.AutoSize = true;
-            this.homeEnabledLabel.Font = new System.Drawing.Font("Reem Kufi", 8.999999F);
-            this.homeEnabledLabel.Location = new System.Drawing.Point(10, 7);
+            this.homeEnabledLabel.Font = new System.Drawing.Font("Microsoft Sans Serif", 8.999999F);
+            this.homeEnabledLabel.Location = new System.Drawing.Point(15, 17);
```
Comment (executionByFork) [inline MasterForm.Designer.cs:259]: The labels should still be centered over the icons and in line with the modpack name. This is probably a product of using the tab control in that everything is shifted, so there's probably not an issue here. I seem to recall that none of the alignment looked off in your screenshots, I'd just like you to double check this for all of the tab pages

**Review by executionByFork — CHANGES_REQUESTED** [review]
Please see my comments and update where needed

File: MasterForm.Designer.cs (hunk 6 above)
Comment (Darksound9) [reply MasterForm.Designer.cs]: Any shifts in positioning are artifacts of me migrating the elements to tab control (cut and paste in the designer) lets not push this PR until all of them are reverted to avoid messing up your positioning code.

File: MasterForm.Designer.cs (hunk 7 above)
Comment (Darksound9) [reply MasterForm.Designer.cs]: Will revert

File: MasterForm.Designer.cs (hunk 8 above)
Comment (Darksound9) [reply MasterForm.Designer.cs:259]: Will revert

File: Api/Modpacks.cs (hunk 3 above)
Comment (Darksound9) [reply Api/Modpacks.cs]: Gotcha. Sounds good for now, done

File: Api/Modpacks.cs (hunk 2 above)
Comment (Darksound9) [reply Api/Modpacks.cs]: Done

File: Api/Modpacks.cs (hunk 1 above)
Comment (Darksound9) [reply Api/Modpacks.cs]: Sounds good, thanks for that I didn't realize. Done

File: Api/MyMods.cs (hunk 5 above)
Comment (Darksound9) [reply Api/MyMods.cs]: Hrmm. So that behavior was intentional. That's an...intriguing design decision. I modified it because, once I encountered it, I intuitively believed it to be a bug.

> if the light is red and the checkbox is checked, the manager will patch the game 
> if the light is green and the checkbox is unchecked the mod manager should unpatch it

This makes no sense to me. For a first time user, the Patch/Unpatch button appears to apply its Patch/Unpatch functionality to the mods that are currently checked, regardless of if they active or inactive mods. It's a process that's applied on a **per selection** basis.

If I open the manager to a fresh MyMods page, to a blank and completely unchecked screen, I intuitively wouldn't expect the Patch/Unpatch button to do anything. But since you've inverted the behavior for already patched mods, it behaves oppositely from unpatched mods with no visual cues to tell the user it behaves differently--making it apply an operations on a blank page with no selections.

Even the error message you provide lends itself towards the behavior I'm expecting. See the output that's generated when I revert this change:
Capture

From what you're telling me, the above is expected behavior.
I think if you want Patch/Unpatch to behave differently for patched vs unpatched mods, the UX for one of the two operations needs to change. Identical UX should result in identical behavior.

> Currently when the mod manager loads, it auto checks any modpack which is enabled.

I see why you were doing this for convenience, but don't you think the new Select Enabled bulk checkbox does this now?

> This also creates a toggling issue, where clicking the button twice because the app is slow could cause the manager to install and then immediately uninstall all mods.

We should probably block input to that button until the operation is complete, regardless of how the checkboxes work.

File: Api/MyMods.cs (hunk 4 above)
Comment (Darksound9) [reply Api/MyMods.cs]: Sure thing. I made that in like, 10 mins. Done

File: Api/MyMods.cs (hunk 5 above)
Comment (executionByFork) [reply Api/MyMods.cs]: Yes this is intended behavior. We could perhaps change that popup you screenshot to instead read "You did not MAKE any changes to the selection", instead of "You did not SELECT any changes" to be more clear. But this is how the mod manager has worked since inception and I haven't received any complaints about it. I think it is convenient in the sense that after pressing the button, any mod that is checked ends up installed in the game, and any mod that is unchecked ends up as not installed in the game, regardless of the current state of your mods installations. Also, given the same UI state that button _does_ perform the same action each time. You aren't taking into account the state of the red/green button, which is a part of the UI.

As for the toggle issue, I was more referring to someone spamming the button while the process finishes, returning control to the UI while the user is spamming, causing them to click the button again and unpatch everything.

Really, either control scheme works. So I think it's best that we don't make a change that is mostly based on opinion, especially since the userbase is already used to the way it works now.

File: MasterForm.Designer.cs (hunk 7 above)
Comment (Darksound9) [reply MasterForm.Designer.cs]: Removed. The tabindex was actually already there, the designer just automatically moved it.

File: MasterForm.Designer.cs (hunk 8 above)
Comment (Darksound9) [reply MasterForm.Designer.cs:259]: This is correct. After reviewing, everything on the tab controls is shifted by a mere 3 pixels. I reverted the top bar, but once I saw that everything was shifted, I opted not to do it by hand. Lots of tedium

It's not noticeable, and I don't believe it will affect your location code as it behaves relative to other items. And if those items are also shifted by 3 pixels, then there's no problem.

I can send you screenshots of how it looks, or you can pull this branch and look for yourself.

File: MasterForm.Designer.cs (hunk 6 above)
Comment (Darksound9) [reply MasterForm.Designer.cs]: See the below comment.

File: Api/MyMods.cs (hunk 5 above)
Comment (Darksound9) [reply Api/MyMods.cs]: >after pressing the button, any mod that is checked ends up installed in the game, and any mod that is unchecked ends up as not installed in the game, regardless of the current state of your mods installations.

I checked back in previous versions, as the version I'd been using for so long doesn't behave this way. Turns out I'd changed this on my personal branch a long time ago and forgotten about it.
I can understand this line of thinking--you're aligning the checkbox selection with installation. Checked means installed, unchecked means not installed, and patch unpatch will _align_ the modpacks to that state.
As opposed to my line of thinking, where the selection applies not to that state, but directly to the patch/unpatch functionality. Whatever is selected will apply patch/unpatch, regardless of the state of the mod.

I'm genuinely amazed that you haven't gotten a complaint up to this point. While your approach is most certainly valid, The current UI implicitly lends itself towards my implementation, between the error message and the column name. I strongly recommend making some changes to make it obvious.
Instead of calling the column name 'Select', it could be called 'Mark for install/uninstall' as the user isn't selecting a mod to run patch/unpatch, they're marking what state the pack should be in. Then the button could be called 'Align' to align the state of the mods with what's marked. Or 'Match Selections'. I still don't love this but at least it's then clear. I don't want users to encounter this and believe it to be a bug rather than an intentional design decision.

Since both approaches are similar, we could add another checkbox to the top right of the page to flip between the two types of control modes. If it's intuitive to me, it will likely be intuitive to many others.

> As for the toggle issue, I was more referring to someone spamming the button while the process finishes, returning control to the UI while the user is spamming, causing them to click the button again and unpatch everything.

...Which is exactly why I said:

> We should probably block input to that button until the operation is complete, regardless of how the checkboxes work.

This should stop people from spamming patch/unpatch operations.

I'll revert this. (Though you can bet I'm keeping it in my personal build ;P )

Comment (Darksound9) [reply Api/MyMods.cs]: > this is how the mod manager has worked since inception and I haven't received any complaints about it.

Not believing this to be possible, I checked in older versions and found that I'd already made this change in my _personal_ version of 0.5, and forgotten that I'd done so. All you really have to do is delete the one !.

That said, I understand your line of thinking--you're making the checkbox selection 1-to-1 with the state of being installed vs not installed, regardless of the state the mod is already in. This approach is certainly valid, however I think your current UI _greatly_ implies my implementation is being performed, between the error message and the column name. I strongly recommend that you make UI tweaks to make your control scheme more obvious. Without doing so users will absolutely make the same mistake, and believe it to be a bug rather than in intentional design decision. Users thinking opposite you already have probably fiddled with it enough to get it to work, and begrudgingly accepted it rather than report it.

You could change the column name to 'Mark as installed'. Then change the Patch/Unpatch text could change to 'Align' or 'Match Selections' to make the mods match the checkboxes. I don't love this, but makes your control scheme apparent. I think the difficulty in making that approach clearer is a testament to the cumbersome nature it has.

Alternatively, we could add a checkbox to flip between the two types of control schemes. It would be easy to implement. Even then I would say each would need to use different types of labels.

> As for the toggle issue, I was more referring to someone spamming the button while the process finishes, returning control to the UI while the user is spamming, causing them to click the button again and unpatch everything.

...that's why I said:

> We should probably block input to that button until the operation is complete, regardless of how the checkboxes work.

I've reverted the checkbox logic. But you better believe I'll be keeping it in my personal branch ;P
